File: packages/hydris/src/http/ProxyResolver.ts
import { Header } from '../constants';
import type { Request } from './Request';

export class ProxyResolver {
    private static readonly FORWARDED_IP_HEADERS = [
        Header.CfConnectingIp,
        Header.TrueClientIp,
        Header.XForwardedFor,
    ];

    static resolveIp(
        req: Request,
        socketIp: string,
        trustProxy: boolean,
    ): string {
        if (trustProxy) {
            for (const header of ProxyResolver.FORWARDED_IP_HEADERS) {
                const value = req.headers.get(header);
                if (value) return value.split(',')[0].trim();
            }
        }
        return socketIp;
    }

    static resolveProtocol(req: Request, trustProxy: boolean): string {
        if (trustProxy) {
            const forwarded = req.headers.get(Header.XForwardedProto);
            if (forwarded) return forwarded.split(',')[0].trim();
        }
        return new URL(req.url).protocol.replace(':', '');
    }
}

File: packages/hydris-sharp/Hydris/Http/ProxyResolver.cs
namespace Hydris.Http;

public static class ProxyResolver {
    private static readonly string[] ForwardedIpHeaders =
        ["CF-Connecting-IP", "True-Client-IP", "X-Forwarded-For"];

    public static string ResolveIp(Request request, string socketIp, bool trustProxy) {
        if (trustProxy) {
            foreach (var header in ForwardedIpHeaders) {
                var value = request.Header(header);
                if (!string.IsNullOrEmpty(value))
                    return First(value);
            }
        }

        return socketIp;
    }

    public static string ResolveProtocol(Request request, string scheme, bool trustProxy) {
        if (trustProxy) {
            var forwarded = request.Header("X-Forwarded-Proto");
            if (!string.IsNullOrEmpty(forwarded))
                return First(forwarded);
        }

        return scheme;
    }

    private static string First(string value) {
        var comma = value.IndexOf(',');
        return (comma >= 0 ? value[..comma] : value).Trim();
    }
}
